fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/SubspaceProviderByKeySpacePath.java

```diff
@@ -70,8 +75,19 @@ public LogMessageKeys logKey() {
         return LogMessageKeys.KEY_SPACE_PATH;
     }
 
+    @Override
+    public String toString(@Nonnull FDBRecordContext context) {
+        CompletableFuture<Subspace> subspaceFuture = databases.get(context.getDatabase().getClusterFile());
+        if (subspaceFuture != null && subspaceFuture.isDone()) {
```

Comment:
`subspaceFuture` might be completed exceptionally; I think we should ignore it in that case.

Reply:
I was about to say that this will never be completed exceptionally because it is never cached if it failed, but reviewing the getSubspaceAsync, it will cache an excepted future.  I think this method is fine, but getSubspceAsync needs to change

Reply:
If you aren't going to cache / share an incomplete future, then the cache should have `Subspace` instead, right?

Reply:
Sorry, I was confused -- I had been looking at the `getSubspaceAsync` from `FDBRecordStore` and not the one locally in this file...ignore my previous comment.

Reply:
subspaceFuture there is a local variable

Reply:
we are caching in whenComplete only when we get a subspace != null (only possible if no exception). so seems a non-issue but correct me if I am mistaken.

Reply:
Well, it seems like it would be cleaner to only cache if `error == null` (which is to say, when there is no exception), but it also seems like either using an `AsyncLoadingCache` so that multiple people attempting to resolve the same keyspace share the same future or caching subspaces instead of subspace futures like @MMcM suggested would make more sense.